batch/batches/Convert/Engines/KJobConversionEngine.class.php

```diff
@@ -146,15 +148,116 @@ public function convertJob ( KalturaConvertJobData &$data )
 		
 		return array ( true , $error_message );// indicate all was converted properly
 	}
-	
+
+	protected function isConversionProgressing($currentModificationTime)
+	{
+		if (kFile::checkFileExists($this->inFilePath))
+		{
+			$newModificationTime = kFile::getFileLastUpdatedTime($this->inFilePath);
+			if ($newModificationTime !== false && $newModificationTime > $currentModificationTime)
+			{
+				return $newModificationTime;
+			}
+		}
+		return false;
+	}
+
+	protected function getReturnValues($handle)
+	{
+		$return_var = 1;
+		$output = false;
+		if ($handle)
+		{
+			$return_var = 0;
+			$file = $this->outFilePath;
+			if (kFile::checkFileExists($file))
+			{
+				$output = kFile::getLineFromFileTail($file , 1);
+			}
+		}
+		return array($output, $return_var);
+	}
+
 	/**
 	 *
 	 */
-	protected function execute_conversion_cmdline($command, &$return_var)
+	protected function execute_conversion_cmdline($command, &$return_var, $jobId = null)
 	{
-		$output = system($command, $return_var);
+		if (isset(KBatchBase::$taskConfig->params->usingSmartJobTimeout) && KBatchBase::$taskConfig->params->usingSmartJobTimeout == 1)
+		{
+			return $this->executeConversionCmdlineSmartTimeout($command, $return_var, $jobId);
+		}
+		else
+		{
+			$output = system($command, $return_var);
+			return $output;
+		}
+	}
+
+	protected function executeConversionCmdlineSmartTimeout($command, &$return_var, $jobId = null)
+	{
+		$handle = popen($command, 'r');
+		stream_set_blocking ($handle,0) ;
+		$currentModificationTime = 0;
+		$lastTimeOutSet = time();
+		$maximumExecutionTime = KBatchBase::$taskConfig->maximumExecutionTime;
+		$extendTime = $maximumExecutionTime ? ($maximumExecutionTime / 3) : dateUtils::HOUR;
+		$timeout = $maximumExecutionTime;
+		while(!feof($handle))
+		{
+			clearstatcache();
+			$buffer = fread($handle,1);
+			$newModificationTime = $this->isConversionProgressing($currentModificationTime);
+			if($newModificationTime)
+			{
+				if ($lastTimeOutSet + $extendTime < time())
+				{
+					list($timeout, $lastTimeOutSet) = self::extendExpiration($jobId, $maximumExecutionTime, $timeout);
```

Comment:
Since lastTimeOutSet is always time(), no need to return it from the function extendExpiration().
Just set it after calling, using a tupple in function return value is not needed here.